consensus/src/quorum_store/tests/batch_aggregator_test.rs

```diff
@@ -0,0 +1,301 @@
+            .unwrap_err(),
+        BatchAggregatorError::OutdatedFragment
+    );
+
+    // Starting with (3,0) should work for a newly created aggregator.
+    aggregator = BatchAggregator::new(1000);
+    assert_ok!(aggregator.append_transactions(3, 0, Vec::new()));
+
+    check_outdated_fragments(&mut aggregator, vec![(3, 0), (2, 0), (2, 7), (0, 0)]);
+    assert_matches!(aggregator.end_batch(3, 1, Vec::new()), Ok((0, _, _)));
+}
+
+#[test]
+fn test_batch_aggregator() {
+    let num_txns = 10;
+    let all_txns = create_vec_serialized_transactions(num_txns);
+    let all_txns_clone = all_txns.clone();
+    let txns_size = all_txns[0].len() * (num_txns as usize);
+    let fragmented_txns1 = split_vec(&all_txns, 5);
+    let fragmented_txns2 = split_vec(&all_txns, 3);
+
+    let mut base_state = IncrementalBatchState::new(txns_size);
+    assert_ok!(base_state.append_transactions(all_txns));
+    let base_res = base_state.finalize_batch().unwrap();
+    assert_eq!(base_res.0, txns_size);
+
+    let mut aggregator = BatchAggregator::new(txns_size);
+    assert_ok!(aggregator.append_transactions(4, 0, fragmented_txns1[0].clone()));
+    assert_ok!(aggregator.append_transactions(4, 1, fragmented_txns1[1].clone()));
+    check_outdated_fragments(&mut aggregator, vec![(4, 0), (4, 1), (3, 0)]);
+    assert_ok_eq!(aggregator.end_batch(4, 2, Vec::new()), base_res);
+
+    assert_ok!(aggregator.append_transactions(6, 0, fragmented_txns1[0].clone()));
+    assert_ok!(aggregator.append_transactions(6, 1, Vec::new()));
+    assert_ok!(aggregator.append_transactions(6, 2, Vec::new()));
+    assert_ok_eq!(
+        aggregator.end_batch(6, 3, fragmented_txns1[1].clone()),
+        base_res
+    );
+
+    assert_eq!(
+        aggregator
+            .append_transactions(6, 0, Vec::new())
+            .unwrap_err(),
+        BatchAggregatorError::OutdatedFragment
+    );
+
+    assert_eq!(
+        aggregator
+            .append_transactions(7, 1, fragmented_txns1[0].clone())
+            .unwrap_err(),
+        BatchAggregatorError::MissedFragment
+    );
+    assert_eq!(
+        aggregator
+            .append_transactions(7, 0, Vec::new())
+            .unwrap_err(),
+        BatchAggregatorError::OutdatedFragment
+    );
+    assert_ok_eq!(aggregator.end_batch(8, 0, all_txns_clone), base_res);
+
+    assert_ok!(aggregator.append_transactions(9, 0, fragmented_txns1[0].clone()));
+    assert_ok!(aggregator.append_transactions(9, 1, fragmented_txns1[0].clone()));
+    assert_eq!(
+        aggregator
+            .append_transactions(9, 2, fragmented_txns1[1].clone())
+            .unwrap_err(),
+        BatchAggregatorError::SizeLimitExceeded
+    );
+
+    assert_eq!(
+        aggregator.end_batch(9, 2, Vec::new()).unwrap_err(),
+        BatchAggregatorError::OutdatedFragment
+    );
+    assert_eq!(
+        aggregator.end_batch(9, 3, Vec::new()).unwrap_err(),
+        BatchAggregatorError::SizeLimitExceeded
+    );
```

Comment:
I think what happened here is OutdatedFragment took precedence, i.e. we detected it was useless and ignored, so (9,2) didn't really end the batch. If it did, then (9,3) would become outdated, as we would then be expecting something from batch 10 (or beyond) next.

LMK if that makes sense or if you'd find a different handling more intuitive and we can discuss.

Interface-wise, we can receive any messages, and since it's from a peer, they might even forge weird messages (e.g. end batch and another end batch message) and we should just have a way of not crashing and processing these error conditions in a reasonable manner. The approach in batch aggregator is to have state and expect smt, and ignore (error/log don't crash) weird messages/calls that don't match what's expected.

Reply:
I understand now. We can keep it this way. Can we just add tests for when only one fragment is sent per batch? 

like `end_batch(10, 0)` and `end_batch(11, 0)`?

Reply:
It did check (10, 0) right after, also added (11, 0).

Reply:
oh it was append, not end, so will add end 11 and add 12